app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostFragment.kt

```diff
@@ -132,9 +137,49 @@ class ZimHostFragment : BaseFragment(), ZimHostCallbacks, ZimHostContract.View {
       }
     }
 
-    startServerButton.setOnClickListener { startStopServer() }
+    startServerButton.setOnClickListener {
+      if (Build.VERSION.SDK_INT > Build.VERSION_CODES.P) {
+        if (checkCoarseLocationAccessPermission()) {
+          startStopServer()
+        }
+      } else {
+        startStopServer()
+      }
+    }
   }
 
+  private fun checkCoarseLocationAccessPermission(): Boolean =
+    if (ContextCompat.checkSelfPermission(
+        requireActivity(),
+        Manifest.permission.ACCESS_COARSE_LOCATION
+      ) == PackageManager.PERMISSION_DENIED
+    ) {
+      when {
+        ActivityCompat.shouldShowRequestPermissionRationale(
+          requireActivity(),
+          Manifest.permission.ACCESS_COARSE_LOCATION
+        ) -> {
+          alertDialogShower.show(
+            KiwixDialog.LocationPermissionRationaleOnHostZimFile,
+            {
+              ActivityCompat.requestPermissions(
+                requireActivity(), arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION),
+                PERMISSION_REQUEST_CODE_COARSE_LOCATION
+              )
+            })
+        }
+        else -> {
+          ActivityCompat.requestPermissions(
+            requireActivity(), arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION),
+            PERMISSION_REQUEST_CODE_COARSE_LOCATION
+          )
+        }
+      }
+      false
+    } else {
+      true
+    }
+
```

Comment:
There is no code to continue the action we were doing previously once the permission is granted, the user will have to press the button after consenting to coarse location

Reply:
@macgills is this right? or do we have an extension function for this?
```
  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<out String>,
    grantResults: IntArray
  ) {
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
      if (requestCode == PERMISSION_REQUEST_CODE_COARSE_LOCATION) {
        startStopServer()
      }
    }
  }
```

Reply:
I think I made a ticket an eon ago to make a class to handle all our permission requests because they are so ugly, I don't think anything ever came of it.

You can see an example [here](https://wajahatkarim.com/2020/05/activity-results-api-onactivityresult/) of the new results api if you scroll down a bit It is... a bit better?  Permissions are still kind of a nightmare to work with

Reply:
>I think I made a ticket an eon ago to make a class to handle all our permission requests 

I can't find it.

Should I add libraries here or create another issue?

Reply:
You can create another issue, I don't care to postpone release much longer